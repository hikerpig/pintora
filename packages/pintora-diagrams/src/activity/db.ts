import { makeIdCounter } from '@pintora/core'
import { BaseDb } from '../util/base-db'
import { BaseDiagramIR } from '../util/ir'
import { OverrideConfigAction, ParamAction, SetTitleAction } from '../util/config'
import { dedent } from '../util/text'

export type Action = {
  id: string
  actionType: string
  message: string
}

export type Condition = {
  id: string
  message: string
  then: {
    label: string
    children: Step[]
  }
  else?: {
    label: string
    children: Step[]
  }
}

export type While = {
  id: string
  message: string
  children: Step[]
  confirmLabel: string | undefined
  denyLabel: string | undefined
}

export type Repeat = {
  id: string
  message: string
  confirmLabel: string | undefined
  firstAction: Action | undefined
  denyLabel: string | undefined
  children: Step[]
}

export type Keyword = {
  id: string
  label: string
}

/** element group */
export type AGroup = {
  id: string
  groupType: string
  name: string
  label?: string
  children: Step[]
  parent?: string
  background: string | null
}

export type Switch = {
  id: string
  message: string | undefined
  children: Step[]
}

export type Case = {
  id: string
  confirmLabel: string | undefined
  children: Step[]
}

export type Fork = {
  id: string
  branches: Step[]
  shouldMerge: boolean
}

export type ForkBranch = {
  id: string
  children: Step[]
}
export type Note = {
  id: string
  text: string
  placement: string
  target?: string
}

export type ArrowLabel = {
  id: string
  text: string
  target?: string
}

type StepValue = Action | Condition | While | Repeat | Keyword | AGroup | Switch | Case | Fork | ForkBranch | ArrowLabel

export type Step<T extends StepValue = StepValue> = {
  type: string
  parentId?: string
  value: T
}

export type ActivityDiagramIR = BaseDiagramIR & {
  steps: Step[]
  notes: Note[]
  arrowLabels: ArrowLabel[]
}

export type ApplyPart =
  | ParamAction
  | OverrideConfigAction
  | {
      type: 'addAction'
      action: Action
    }
  | {
      type: 'condition'
      message: string
      then: {
        label: string
        children: ApplyPart[]
      }
      else?: {
        label: string
        children: ApplyPart[]
      }
    }
  | {
      type: 'while'
      message: string
      confirmLabel: string
      denyLabel: string
      children: ApplyPart[]
    }
  | {
      type: 'repeat'
      message: string
      confirmLabel: string
      denyLabel: string
      firstAction: Action
      children: ApplyPart[]
    }
  | {
      type: 'switch'
      message: string
      children: ApplyPart[]
    }
  | {
      type: 'case'
      confirmLabel: string
      children: ApplyPart[]
    }
  | {
      type: 'keyword'
      label: string
    }
  | {
      type: 'group'
      groupType: string
      name: string
      label?: string
      children: ApplyPart[]
      background: string | null
    }
  | {
      type: 'note'
      text: string
      placement: string
    }
  | {
      type: 'arrowLabel'
      text: string
    }
  | {
      type: 'fork'
      branches: ApplyPart[]
      shouldMerge: boolean
    }
  | {
      type: 'forkBranch'
      children: ApplyPart[]
    }
  | SetTitleAction

type DbApplyState = {
  prevStepId?: string | undefined
  parentId?: string | undefined
}

class ActivityDb extends BaseDb {
  protected steps: Step[] = []
  protected notes: Note[] = []
  protected arrowLabels: ArrowLabel[] = []
  protected idCounter = makeIdCounter()

  protected makeId() {
    return this.idCounter.next()
  }

  apply(part: ApplyPart, ignoreAdd?: boolean, state?: DbApplyState): Step
  apply(part: ApplyPart[], ignoreAdd?: boolean, state?: DbApplyState): Step[]
  apply(part: ApplyPart | ApplyPart[], ignoreAdd = false, state: DbApplyState = {}): Step | Step[] {
    if (Array.isArray(part)) {
      const partResults: any[] = []
      let currentStep: Step
      part.forEach(p => {
        const result = this.apply(p, ignoreAdd, { ...state, prevStepId: currentStep?.value?.id })
        if (result) {
          partResults.push(result)
          currentStep = result
        }
      })
      return partResults
    }
    if (!part) return
    // console.log('[activity] apply', part)
    let step: Step
    switch (part.type) {
      case 'addAction':
        {
          const action = { ...part.action, id: this.makeId() }
          step = { type: 'action', value: action }
        }
        break
      case 'condition':
        {
          const id = this.makeId()
          const thenResult = this.apply(part.then.children, true, { ...state, parentId: id })
          const condition: Condition = {
            id,
            message: part.message,
            then: {
              label: part.then.label,
              children: thenResult,
            },
          }
          if (part.else) {
            const elseResult = this.apply(part.else.children, true, { ...state, parentId: id })
            condition.else = {
              label: part.else.label,
              children: elseResult,
            }
          }
          step = { type: 'condition', value: condition }
        }
        break
      case 'while': {
        const id = this.makeId()
        const loopResult = this.apply(part.children, true, { ...state, parentId: id })
        const whileSentence: While = {
          id,
          message: part.message,
          children: loopResult,
          confirmLabel: part.confirmLabel,
          denyLabel: part.denyLabel,
        }
        step = { type: 'while', value: whileSentence }
        break
      }
      case 'repeat': {
        const id = this.makeId()
        const loopResult = this.apply(part.children, true, { ...state, parentId: id })
        const sentence: Repeat = {
          id,
          message: part.message,
          children: loopResult,
          confirmLabel: part.confirmLabel,
          denyLabel: part.denyLabel,
          firstAction: part.firstAction,
        }
        step = { type: 'repeat', value: sentence }
        break
      }
      case 'switch': {
        const id = this.makeId()
        const cases = this.apply(part.children, true, { ...state, parentId: id })
        const switchSentence: Switch = {
          id,
          message: part.message,
          children: cases as any,
        }
        step = { type: 'switch', value: switchSentence }
        break
      }
      case 'case': {
        const id = this.makeId()
        const children = this.apply(part.children, true, { ...state, parentId: id })
        const caseClause: Case = {
          id,
          confirmLabel: part.confirmLabel,
          children,
        }
        step = { type: 'case', value: caseClause }
        break
      }
      case 'fork': {
        const id = this.makeId()
        const branches = this.apply(part.branches, true, { ...state, parentId: id })
        const forkSentence: Fork = {
          id,
          shouldMerge: part.shouldMerge,
          branches,
        }
        step = { type: 'fork', value: forkSentence }
        break
      }
      case 'forkBranch': {
        const id = this.makeId()
        const children = this.apply(part.children, true, { ...state, parentId: id })
        const forkBranch: ForkBranch = {
          id,
          children,
        }
        step = { type: 'forkBranch', value: forkBranch }
        break
      }
      case 'keyword': {
        step = { type: 'keyword', value: { id: this.makeId(), label: part.label } }
        break
      }
      case 'group': {
        const id = this.makeId()
        const childrenResult = this.apply(part.children, true, { ...state, parentId: id })
        step = { type: 'group', value: { id, ...part, children: childrenResult } }
        break
      }
      case 'note': {
        const value: Note = { id: this.makeId(), ...part, text: dedent(part.text) }
        const prevStepId = state.prevStepId
        if (prevStepId) {
          value.target = prevStepId
        }
        this.notes.push(value)
        break
      }
      case 'arrowLabel': {
        const value: ArrowLabel = { id: this.makeId(), ...part }
        const prevStepId = state.prevStepId
        if (prevStepId) {
          value.target = prevStepId
        }
        this.arrowLabels.push(value)
        break
      }
      case 'setTitle': {
        this.title = part.text
        break
      }
      case 'addParam': {
        this.configParams.push(part)
        break
      }
      case 'overrideConfig': {
        this.addOverrideConfig(part)
        break
      }
      default: {
      }
    }
    if (step && !ignoreAdd) {
      this.steps.push(step)
    }
    if (step) {
      step.parentId = state.parentId
    }
    return step as any
  }

  getDiagramIR(): ActivityDiagramIR {
    return {
      steps: this.steps,
      notes: this.notes,
      arrowLabels: this.arrowLabels,
      ...this.getBaseDiagramIR(),
    }
  }

  clear() {
    super.clear()
    this.idCounter.reset()
    this.steps = []
    this.notes = []
  }
}

export const db = new ActivityDb()
