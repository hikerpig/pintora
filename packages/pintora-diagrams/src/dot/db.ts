/* eslint-disable @typescript-eslint/no-empty-function */
import { BaseDiagramIR } from '../util/ir'
import { OverrideConfigAction } from '../util/config'
import { ActionHandler, BaseDb, MakeAction } from '../util/base-db'
import { ConfigParam } from '@pintora/core'

type DOTActionPayloads = {
  overrideConfig: OverrideConfigAction
  addParam: ConfigParam
  addGraph: {
    graph: ParserDOTGraph
  }
}

export type Action = MakeAction<DOTActionPayloads>

export type DotIR = BaseDiagramIR & {
  graph: DOTGraph
}

export class DOTDb extends BaseDb {
  graph: DOTGraph

  ACTION_HANDLERS: { [K in keyof DOTActionPayloads]: ActionHandler<DOTActionPayloads, DOTDb, K> } = {
    overrideConfig(action) {
      this.addOverrideConfig(action)
    },
    addParam(action) {
      this.configParams.push(action)
    },
    addGraph(action) {
      function attrListToObject<R = any>(list: Attr[]) {
        return list.reduce((acc, attr) => {
          acc[attr.id] = attr.eq
          return acc
        }, {}) as R
      }

      function processGraph<T extends DOTGraph | Subgraph>(graph: T): T {
        const newGraph = { ...graph } as unknown as T
        const attrList: Attr[] = []
        const childList: T['children'] = []
        const attrsCollection: AttrsCollection = {}
        for (const child of newGraph.children) {
          let newChild = child
          switch (child.type) {
            case 'attr_stmt': {
              const _child = child as unknown as ParserAttrStmt
              const target = child.target || 'graph'
              if ('attr_list' in _child) {
                attrsCollection[target] = Object.assign(
                  attrsCollection[target] || {},
                  attrListToObject(_child.attr_list),
                )
                newChild = null
              }
              break
            }
            case 'node_stmt':
            case 'edge_stmt': {
              const _child = child as unknown as ParserEntityStmt
              if ('attr_list' in _child) {
                child.attrs = attrListToObject(_child.attr_list) as any
                delete _child.attr_list
              }
              break
            }
            case 'subgraph': {
              newChild = processGraph(child)
              break
            }
            case 'attr': {
              attrList.push(child)
              newChild = null
              break
            }
          }
          if (newChild) childList.push(newChild)
        }
        newGraph.children = childList
        if (attrList.length) {
          attrsCollection.graph = Object.assign(attrsCollection.graph || {}, attrListToObject(attrList))
        }
        if (Object.keys(attrsCollection).length) newGraph.attrs = attrsCollection
        return newGraph
      }
      const newGraph = processGraph<DOTGraph>(action.graph as any)

      this.graph = newGraph
    },
  }

  getDiagramIR(): DotIR {
    this.processAttrs()
    return {
      ...super.getBaseDiagramIR(),
      graph: this.graph,
    }
  }

  processAttrs() {}

  apply(action: Action | Action[]) {
    if (!action) return
    if (Array.isArray(action)) {
      action.forEach(a => this.apply(a))
      return
    }
    if (action.type in this.ACTION_HANDLERS) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.ACTION_HANDLERS[action.type].call(this, action as any)
    }
  }

  override clear() {
    super.clear()
    this.graph = undefined
  }
}

const db = new DOTDb()

export default db

// data compatible with dotparser
export type CompassPt = 'n' | 'ne' | 'e' | 'se' | 's' | 'sw' | 'w' | 'nw'
export type Stmt = AttrStmt | EdgeStmt | NodeStmt | Subgraph | Attr

/** location or position */
// export type Loc = 't' | 'b'

export interface Port {
  type: 'port'
  id: string | number
  compass_pt?: CompassPt
}

export interface NodeId {
  type: 'node_id'
  id: string
  port?: Port
}

export interface HTMLString {
  type: 'id'
  value: 'string'
  html: true
}

// https://graphviz.org/doc/info/attrs.html
export type NodeAttrs = {
  label?: string
  /** stroke color */
  color?: string
  /** label text color */
  fontcolor?: string
  /** background color */
  bgcolor?: string
  /** font family for node label */
  fontname?: string
  /** font size for node label */
  fontsize?: number
  /** [pintora specific], margin top */
  margint?: number
  /** [pintora specific], margin bottom */
  margintb?: number
  /** [pintora specific], margin left */
  marginl?: number
  /** [pintora specific], margin right */
  marginr?: number
}

export type EdgeStyle = 'invis'

export type DOTArrowType = 'normal' | 'box' | 'obox' | 'dot' | 'odot' | 'open' | 'diamond' | 'ediamond'

export type EdgeAttrs = {
  label?: string
  /** edge line color */
  color?: string
  /** edge label text color */
  fontcolor?: string
  /** edge style */
  style?: EdgeStyle
  /** edge line width */
  penwidth?: number
  /** font family for edge label */
  fontname?: string
  /** font size for edge label */
  fontsize?: number
  /** arrow type for arrow head */
  arrowhead?: DOTArrowType
}

export type GraphAttrs = {
  label?: string
  /** stroke color */
  color?: string
  /** background color */
  bgcolor?: string
  /** font family for graph label */
  fontname?: string
  /** font size for graph label */
  fontsize?: number
}

export type AttrsCollection = Partial<{
  graph: GraphAttrs
  node: NodeAttrs
  edge: EdgeAttrs
}>

export interface Attr {
  type: 'attr'
  id: string | number
  eq: string | HTMLString
}

export interface Subgraph {
  type: 'subgraph'
  children: Stmt[]
  id?: string
  attrs?: AttrsCollection
}

export interface AttrStmt<T = NodeAttrs> {
  type: 'attr_stmt'
  target: 'graph' | 'node' | 'edge'
  attrs?: T
}

export interface EdgeStmt<T = EdgeAttrs> {
  type: 'edge_stmt'
  edge_list: (Subgraph | NodeId)[]
  attrs?: T
}

export interface NodeStmt<T = NodeAttrs> {
  type: 'node_stmt'
  nodeId: NodeId
  attrs?: T
}

export interface DOTGraph {
  type: 'graph' | 'digraph'
  children: Stmt[]
  id: string
  attrs?: AttrsCollection
}

type ToParserEntityStmt<T> = Omit<T, 'attrs'> & {
  attr_list?: Attr[]
}

export type ParserEdgeStmt = ToParserEntityStmt<EdgeStmt>
export type ParserNodeStmt = ToParserEntityStmt<NodeStmt>
export type ParserSubgraph = ToParserEntityStmt<GraphAttrs>

export type ParserEntityStmt = ParserEdgeStmt | ParserNodeStmt | ParserSubgraph

export type ParserAttrStmt = ToParserEntityStmt<AttrStmt>

export type ParserDOTGraph = Omit<DOTGraph, 'children'> & {
  children: ParserEntityStmt[]
}
