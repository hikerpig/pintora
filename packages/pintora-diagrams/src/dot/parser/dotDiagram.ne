# https://graphviz.org/doc/info/lang.html
@preprocessor typescript
@lexer lexer
@skip_unmatch %WS
@include "whitespace.ne"
@include "config.ne"
@include "comment.ne"

@{%
import * as moo from '@hikerpig/moo'
import {
  tv,
  textToCaseInsensitiveRegex,
  VALID_TEXT_REGEXP,
  COMMENT_LINE_REGEXP,
  QUOTED_WORD_REGEXP,
  configLexerMainState,
  configLexerconfigStatementState,
  L_PAREN_REGEXP,
  R_PAREN_REGEXP,
  COLOR_REGEXP,
  MOO_NEWLINE,
  getQuotedWord,
} from '../../util/parser-shared'
import type { Action, ParserDOTGraph, Attr, ParserEntityStmt, NodeId, Subgraph } from '../db'

const COMMON_TOKEN_RULES = {
  VALID_TEXT: { match: VALID_TEXT_REGEXP, fallback: true },
}

let lexer = moo.states({
  main: {
    NL: MOO_NEWLINE,
    WS: { match: /[ \t]+/, lineBreaks: false },
    QUOTED_WORD: QUOTED_WORD_REGEXP,
    COLOR: COLOR_REGEXP,
    SEMICOLON: /;/,
    COLON: /:/,
    COMMA: /,/,
    DOT_DIAGRAM: /dotDiagram/,
    L_PAREN: L_PAREN_REGEXP,
    R_PAREN: R_PAREN_REGEXP,
    L_BRACKET: { match: /\{/ },
    R_BRACKET: { match: /\}/ },
    L_SQ_BRACKET: { match: /\[/ },
    R_SQ_BRACKET: { match: /\]/ },
    EQ: { match: /=/ },
    SUBGRAPH: { match: /subgraph/ },
    START_NOTE: textToCaseInsensitiveRegex('@note'),
    END_NOTE: textToCaseInsensitiveRegex('@end_note'),
    COMMENT_LINE: COMMENT_LINE_REGEXP,
    DOT_SLASH_COMMENT: /\/\/.*/,
    DOT_BLOCK_COMMENT_START: { match: /\/\*/, push: 'blockComment' },
    ...configLexerMainState,
    VALID_TEXT: { match: VALID_TEXT_REGEXP, fallback: true },
  },
  configStatement: {
    ...configLexerconfigStatementState,
    ...COMMON_TOKEN_RULES,
  },
  blockComment: {
    DOT_BLOCK_COMMENT_END: { match: /\*\//, pop: 1 },
    ANY_COMMENT_TEXT: { match: /(?:\s\w\d\n\r)+(?!\*\/)/, fallback: true },
  }
})

function rNull() {
  return null
}
%}

start -> __ start {% (d) => d[1] %}
	| %DOT_DIAGRAM document {%
      function(d) {
        return d[1]
      }
    %}

document -> null
  | document statementWrap {%
    (d) => {
        // console.log('[doc line]', d[1])
        let r = d[0]
        if (d[1]) {
          r = d[0].concat(d[1])
        }
        return r
      }
    %}

statementWrap ->
	  %WS:* statement {% (d) => {
      // console.log('[line]', JSON.stringify(d[1], null, 2))
      return d[1]
    } %}
	| %WS:? dotCommentSegment %NL {% null %}
	| %WS:? %NL {% null %}

statement ->
    # as 'graph' rule in DOT spec
    ("graph" | "digraph") %VALID_TEXT:? %L_BRACKET %NL:? stmtList %R_BRACKET {%
      function(d) {
        const children = d[4]
        const id = d[1] ? tv(d[1]): ''
        const action: Action = {
          type: 'addGraph',
          graph: {
            type: tv(d[0][0]),
            id,
            children,
          } as ParserDOTGraph
        }
        return action
      }
    %}
  | paramStatement %NL
  | configOpenCloseStatement %NL
  | comment %NL

dotCommentSegment ->
    %DOT_SLASH_COMMENT {% rNull %}
  | %DOT_BLOCK_COMMENT_START %ANY_COMMENT_TEXT:* %DOT_BLOCK_COMMENT_END {% rNull %}

# as 'stmt_list' rule in DOT spec
stmtList ->
    dotStmt (%SEMICOLON | %NL) _ stmtList:? {%
      function(d) {
        const stmtList = d[3] || []
        const stmt = d[0][0]
        if (!stmt) return stmtList
        return [stmt, ...stmtList]
      }
    %}

# as 'stmt' rule in DOT spec
dotStmt ->
    attrStmt
  | nodeStmt
  | edgeStmt
  | singleAttrStmt
  | subgraph
  | dotCommentSegment {% null %}
  | comment {% null %}

# as 'attr_stmt' rule in DOT spec
attrStmt ->
    ("graph" | "node" | "edge") __ attrList {%
      function(d) {
        const target = tv(d[0][0])
        const attr_list = d[2]
        return {
          type: 'attr_stmt',
          target,
          attr_list,
        } as ParserEntityStmt
      }
    %}

singleAttrStmt ->
    %VALID_TEXT %EQ (%VALID_TEXT | %QUOTED_WORD) (%SEMICOLON | %COMMA):? {%
      function(d) {
        const id = tv(d[0])
        const eqValToken = d[2][0]
        const eq = eqValToken.type === 'QUOTED_WORD' ? getQuotedWord(eqValToken): tv(eqValToken)
        const attr: Attr = {
          type: 'attr',
          id,
          eq,
        }
        return attr
      }
    %}

nodeStmt ->
    nodeId attrList:? {%
      function(d) {
        const nodeStmt: ParserEntityStmt = {
          type: 'node_stmt',
          nodeId: d[0],
        }
        if (d[1]) {
          nodeStmt.attr_list = d[1]
        }
        return nodeStmt
      }
    %}
  | nodeId %L_SQ_BRACKET %QUOTED_WORD %R_SQ_BRACKET {%
      function(d) {
        const nodeId = d[0]
        const label = getQuotedWord(d[2])
        return {
          type: 'node_stmt',
          nodeId: d[0],
          attrs: {
            label,
          }
        }
      }
    %}
  | nodeId {%
      function(d) {
        const nodeId = d[0]
        return {
          type: 'node_stmt',
          nodeId: d[0],
        }
      }
    %}

# as 'edge_stmt' rule in DOT spec
edgeStmt ->
    (nodeId) edgeRHS attrList:? {%
      function(d) {
        const startNode = d[0][0]
        const edge_list = [startNode, ...d[1]]
        const edgeStmt: ParserEntityStmt = {
          type: 'edge_stmt',
          edge_list,
        }
        if (d[2]) {
          edgeStmt.attr_list = d[2]
        }
        return edgeStmt
      }
    %}

edgeRHS ->
    edgeop %WS:? (nodeId) edgeRHS:? {%
      function(d) {
        const edgeList = d[3] || []
        const edge = d[2][0]
        return [edge, ...edgeList]
      }
    %}

edgeop -> "->"
  | "--"

attrList ->
    %L_SQ_BRACKET attrItems %R_SQ_BRACKET attrList:? {%
      function(d) {
        const attrList = d[3] ? d[3][0]: []
        const attrItems = d[1]
        return d[1]
      }
    %}

# as 'a_list' rule in DOT spec
attrItems ->
    %VALID_TEXT %EQ (%VALID_TEXT | %QUOTED_WORD) (%SEMICOLON | %COMMA):? attrItems:? {%
      function(d) {
        const attrItems = d[4] ? d[4]: []
        const id = tv(d[0])
        const eqValToken = d[2][0]
        const eq = eqValToken.type === 'QUOTED_WORD' ? getQuotedWord(eqValToken): tv(eqValToken)
        const attr: Attr = {
          type: 'attr',
          id,
          eq,
        }

        return [attr, ...attrItems]
      }
    %}

nodeId ->
    %VALID_TEXT {%
      function(d) {
        return {
          type: 'node_id',
          id: tv(d[0]).trim()
        } as NodeId
      }
    %}
  | %QUOTED_WORD {%
      function(d) {
        return {
          type: 'node_id',
          id: getQuotedWord(d[0]).trim()
        } as NodeId
      }
    %}

# as 'subgraph' rule in DOT spec
subgraph ->
    %SUBGRAPH %VALID_TEXT:? %L_BRACKET _ stmtList _ %R_BRACKET {%
      function(d) {
        const children = d[4]
        const subgraph: Subgraph = {
          type: 'subgraph',
          children,
        }
        if (d[1]) {
          subgraph.id = tv(d[1]).trim()
        }
        return subgraph
      }
    %}
