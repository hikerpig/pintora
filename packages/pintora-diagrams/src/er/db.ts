import { BaseDiagramIR } from '../util/ir'
import { ConfigParam } from '../util/config'
import { BaseDb } from '../util/base-db'

export enum Cardinality {
  ZERO_OR_ONE = 'ZERO_OR_ONE',
  ZERO_OR_MORE = 'ZERO_OR_MORE',
  ONE_OR_MORE = 'ONE_OR_MORE',
  ONLY_ONE = 'ONLY_ONE',
  MORE = 'MORE',
}

export enum Identification {
  NON_IDENTIFYING = 'NON_IDENTIFYING',
  IDENTIFYING = 'IDENTIFYING',
}

export type Attribute = {
  attributeType: string
  attributeName: string
  attributeKey?: string
  comment?: string
}

export type Entity = {
  itemId: string
  attributes: Attribute[]
}

export type Relationship = {
  itemId: string
  entityA: string
  roleA: string
  entityB: string
  relSpec: RelSpec
}

export type RelSpec = {
  cardA: Cardinality
  cardB: Cardinality
  relType: Identification
}

export type Inheritance = {
  sup: string
  sub: string
}

export type ErDiagramIR = BaseDiagramIR & {
  entities: Record<string, Entity>
  relationships: Relationship[]
  inheritances: Inheritance[]
}

export class ErDb extends BaseDb {
  Cardinality = Cardinality
  Identification = Identification

  entities: Record<string, Entity> = {}
  relationships: Relationship[] = []
  inheritances: Inheritance[] = []

  addEntity(name: string) {
    if (!this.entities[name]) {
      this.entities[name] = { attributes: [], itemId: `entity-${name}` }
    }
    return this.entities[name]
  }
  addRelationship(entityA: string, roleA: string, entityB: string, relSpec: RelSpec) {
    const itemId = `relationship-${entityA}-${entityB}`
    const rel: Relationship = {
      entityA,
      roleA,
      entityB,
      relSpec,
      itemId,
    }

    this.relationships.push(rel)
  }
  addInheritance(sup: string, sub: string) {
    this.inheritances.push({ sup, sub })
  }
  getDiagramIR(): ErDiagramIR {
    return {
      ...super.getBaseDiagramIR(),
      entities: this.entities,
      relationships: this.relationships,
      inheritances: this.inheritances,
    }
  }
  addTitle(title: string) {
    this.title = title
  }
  addAttributes(name: string, attributes: Attribute[]) {
    const entity = this.addEntity(name)
    entity.attributes.push(...attributes)
  }
  addParam(styleParam: ConfigParam) {
    this.configParams.push(styleParam)
  }

  override clear() {
    super.clear()
    this.entities = {}
    this.relationships = []
    this.inheritances = []
    this.configParams = []
    this.title = ''
  }
}

const db = new ErDb()

export default db
