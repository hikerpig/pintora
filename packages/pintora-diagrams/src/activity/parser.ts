import { db } from './db'
import grammar from './parser/activityDiagram'
import { genParserWithRules } from '../util/parser-util'

export const parse = genParserWithRules(grammar, {
  dedupeAmbigousResults: true,
  postProcess(results) {
    db.apply(results)
    return results
  },
})
