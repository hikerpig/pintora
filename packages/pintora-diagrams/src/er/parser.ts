import { genParserWithRules } from '../util/parser-util'
import db from './db'
import grammar from './parser/erDiagram'

export const parse = genParserWithRules(grammar, {
  dedupeAmbigousResults: true,
  postProcess(results) {
    db.apply(results)
    return results
  },
})
