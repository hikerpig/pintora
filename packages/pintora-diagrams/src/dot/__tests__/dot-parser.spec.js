import { parse } from '../parser'
import db from '../db'
import { stripStartEmptyLines } from '@pintora/test-shared'

// some test set https://gitlab.com/graphviz/graphviz/-/tree/main/tests/graphs
describe('dot parser', () => {
  afterEach(() => {
    db.clear()
  })

  it('should parse node', () => {
    const example = stripStartEmptyLines(`
dotDiagram

graph ER {
	name [shape=box]; course; student;
}
    `)
    parse(example)
    const ir = db.getDiagramIR()
    // console.log(JSON.stringify(ir, null, 2))
    expect(ir.graph).toMatchSnapshot()
  })

  it('should parse node in shorthand', () => {
    const example = stripStartEmptyLines(`
dotDiagram
graph ER {
	name["a long long name"]
}
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.graph).toMatchSnapshot()
  })

  it('should parse edge', () => {
    const example = stripStartEmptyLines(`
  dotDiagram
  graph Test {
    name0 -- course;
    course -- "C-I" [label="n",len=1.00];
  }
      `)
    parse(example)
    const ir = db.getDiagramIR()
    // console.log(JSON.stringify(ir, null, 2))
    expect(ir.graph).toMatchSnapshot()
  })

  it('should parse directed edge', () => {
    const example = stripStartEmptyLines(`
  dotDiagram
  digraph Test {
    a -> b;
  }
      `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.graph).toMatchSnapshot()
  })

  it('should parse attr stmt', () => {
    const example = stripStartEmptyLines(`
  dotDiagram
  graph {
    color=lightgrey;
    node [color=red, fontcolor=blue]
    edge [color=purple, fontcolor=green]

    subgraph sub {
      label = "process";
    }
  }
      `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.graph).toMatchSnapshot()
  })

  it('should parse subgraph', () => {
    const example = stripStartEmptyLines(`
  dotDiagram
  graph Test {
    subgraph S1 {
      n1;
    }
    n2;

    n1 -- n2;
  }
      `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.graph).toMatchSnapshot()
  })

  it('should parse dot comment', () => {
    const example = stripStartEmptyLines(`
  dotDiagram
  // comment 1
  graph Test {
    // comment 2
    some_node
    %% pintora comment 1
    /**
     * block - comment
     */
  }
  %% pintora comment 2
      `)
    parse(example)
    const ir = db.getDiagramIR()
    // console.log(JSON.stringify(ir, null, 2))
    expect(ir.graph).toMatchSnapshot()
  })
})
