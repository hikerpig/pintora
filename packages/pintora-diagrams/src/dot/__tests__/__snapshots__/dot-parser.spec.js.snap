// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dot parser should parse attr stmt 1`] = `
Object {
  "attrs": Object {
    "edge": Object {
      "color": "purple",
      "fontcolor": "green",
    },
    "graph": Object {
      "color": "lightgrey",
    },
    "node": Object {
      "color": "red",
      "fontcolor": "blue",
    },
  },
  "children": Array [
    Object {
      "attrs": Object {
        "graph": Object {
          "label": "process",
        },
      },
      "children": Array [],
      "id": "sub",
      "type": "subgraph",
    },
  ],
  "id": "",
  "type": "graph",
}
`;

exports[`dot parser should parse directed edge 1`] = `
Object {
  "children": Array [
    Object {
      "edge_list": Array [
        Object {
          "id": "a",
          "type": "node_id",
        },
        Object {
          "id": "b",
          "type": "node_id",
        },
      ],
      "type": "edge_stmt",
    },
  ],
  "id": "Test",
  "type": "digraph",
}
`;

exports[`dot parser should parse dot comment 1`] = `
Object {
  "children": Array [
    Object {
      "nodeId": Object {
        "id": "some_node",
        "type": "node_id",
      },
      "type": "node_stmt",
    },
  ],
  "id": "Test",
  "type": "graph",
}
`;

exports[`dot parser should parse edge 1`] = `
Object {
  "children": Array [
    Object {
      "edge_list": Array [
        Object {
          "id": "name0",
          "type": "node_id",
        },
        Object {
          "id": "course",
          "type": "node_id",
        },
      ],
      "type": "edge_stmt",
    },
    Object {
      "attrs": Object {
        "label": "n",
        "len": "1.00",
      },
      "edge_list": Array [
        Object {
          "id": "course",
          "type": "node_id",
        },
        Object {
          "id": "C-I",
          "type": "node_id",
        },
      ],
      "type": "edge_stmt",
    },
  ],
  "id": "Test",
  "type": "graph",
}
`;

exports[`dot parser should parse node 1`] = `
Object {
  "children": Array [
    Object {
      "attrs": Object {
        "shape": "box",
      },
      "nodeId": Object {
        "id": "name",
        "type": "node_id",
      },
      "type": "node_stmt",
    },
    Object {
      "nodeId": Object {
        "id": "course",
        "type": "node_id",
      },
      "type": "node_stmt",
    },
    Object {
      "nodeId": Object {
        "id": "student",
        "type": "node_id",
      },
      "type": "node_stmt",
    },
  ],
  "id": "ER",
  "type": "graph",
}
`;

exports[`dot parser should parse node in shorthand 1`] = `
Object {
  "children": Array [
    Object {
      "attrs": Object {
        "label": "a long long name",
      },
      "nodeId": Object {
        "id": "name",
        "type": "node_id",
      },
      "type": "node_stmt",
    },
  ],
  "id": "ER",
  "type": "graph",
}
`;

exports[`dot parser should parse subgraph 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "nodeId": Object {
            "id": "n1",
            "type": "node_id",
          },
          "type": "node_stmt",
        },
      ],
      "id": "S1",
      "type": "subgraph",
    },
    Object {
      "nodeId": Object {
        "id": "n2",
        "type": "node_id",
      },
      "type": "node_stmt",
    },
    Object {
      "edge_list": Array [
        Object {
          "id": "n1",
          "type": "node_id",
        },
        Object {
          "id": "n2",
          "type": "node_id",
        },
      ],
      "type": "edge_stmt",
    },
  ],
  "id": "Test",
  "type": "graph",
}
`;
